File: solutions/1800-1899/1896.minimum-cost-to-change-the-final-value-of-expression/Solution.py
from typing import List

class Solution:
    def minOperationsToFlip(self, expression: str) -> int:
        stack = []
        for char in expression:
            if char in ('0', '1', '|', '&'):
                stack.append((char, 1 if char in '01' else 0))
            elif char == '(':
                stack.append(char)
            else:
                e2 = stack.pop()
                stack.pop()
                e1 = stack.pop()
                if e2[0] == '0' and e1[0] == '|':
                    stack.append(('0', 1))
                else:
                    stack.append((max(e1, e2, key=lambda x: (x[0] == '1', -x[1]))))
        while len(stack) > 1:
            e2 = stack.pop()
            op = stack.pop()
            e1 = stack.pop()
            if e2[0] == e1[0]:
                stack.append((e2[0], 1 + min(e1[1], e2[1])))
            else:
                stack.append((e2[0] if op[0] == '|' else e1[0], max(e1[1], e2[1])))
        return stack[0][1]

File: solutions/1800-1899/1896.minimum-cost-to-change-the-final-value-of-expression/Solution.cpp
class Solution {
    stack<vector<int>> num;
    stack<char> op;
    void eval() {
        auto b = num.top(); num.pop();
        auto a = num.top(); num.pop();
        char c = op.top(); op.pop();
        if (c == '&') {
            num.push({
                min({ a[0] + b[0], a[1] + b[0], a[0] + b[1] }),
                min({ a[1] + b[1], a[1] + b[0] + 1, a[0] + b[1] + 1, a[1] + b[1] + 1 })
            });
        } else {
            num.push({
                min({ a[0] + b[0], a[0] + b[1] + 1, a[1] + b[0] + 1, a[0] + b[0] + 1 }),
                min({ a[1] + b[1], a[1] + b[0], a[0] + b[1] })
            });
        }
    }
public:
    int minOperationsToFlip(string s) {
        for (char c : s) {
            if (isdigit(c)) {
                if (c == '0') num.push({0, 1});
                else num.push({1, 0});
            } else if (c == '(') {
                op.push(c);
            } else if (c == ')') {
                while (op.top() != '(') eval();
                op.pop();
            } else {
                while (op.size() && op.top() != '(') eval();
                op.push(c);
            }
        }
        while (op.size()) eval();
        return max(num.top()[0], num.top()[1]);
    }
};
